AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,236 @@
+        
+        // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+        MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+        UIImage *pinImage = pin.image;
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        for (id<MKAnnotation>annotation in _annotations)
+        {
+          CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+          if (CGRectContainsPoint(finalImageRect, point))
+          {
+            CGPoint pinCenterOffset = pin.centerOffset;
+            point.x -= pin.bounds.size.width / 2.0;
+            point.y -= pin.bounds.size.height / 2.0;
+            point.x += pinCenterOffset.x;
+            point.y += pinCenterOffset.y;
+            [pinImage drawAtPoint:point];
+          }
+        }
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        _mapImage.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = _nodeSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapView) {
+    _mapView = [[ASDisplayNode alloc]initWithViewBlock:^UIView *{
+      MKMapView *mapView = [[MKMapView alloc]initWithFrame:CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height)];
+      mapView.delegate = _mapDelegate;
+      [mapView setRegion:_options.region];
+      [mapView addAnnotations:_annotations];
+      return mapView;
+    }];
+    [self addSubnode:_mapView];
+    _mapImage.hidden = YES;
+  }
+}
+
+- (void)removeLiveMap
+{
+  if (_mapView) {
+    [_mapView removeFromSupernode];
+    _mapView = nil;
+    _mapImage.hidden = NO;
+  }
+}
+
+#pragma mark - Layout
+- (CGSize)calculateSizeThatFits:(CGSize)constrainedSize
+{
+  _nodeSize = CGSizeEqualToSize(CGSizeZero, _mapSize) ? CGSizeMake(constrainedSize.width, _options.size.height) : _mapSize;
+  if (_mapImage) {
+    [_mapImage calculateSizeThatFits:_nodeSize];
+  }
+  return _nodeSize;
+}
+
+// Layout isn't usually needed in the box model, but since we are making use of MKMapView which is hidden in an ASDisplayNode this is preferred.
+- (void)layout
+{
+  [super layout];
+  if (_mapView) {
+    MKMapView *mapView = (MKMapView *)_mapView.view;
+    mapView.frame = CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height);
+  }
+  else {
+    _mapImage.frame = CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height);
+    if (!CGSizeEqualToSize(_maxSize, self.bounds.size)) {
+      _mapImage.preferredFrameSize = self.bounds.size;
+      _maxSize = self.bounds.size;
```

Comment:
Yeah totally agree, quite messy this way. I was using them mainly to accurately detect orientation changes. Any ideas on that? Although if we abstract to reloading the image on size changes this will be superseded.